inference: avoid inferring unreachable code methods

vtjnash opened the pull request:
Avoids causing issues in matching_cache_argtypes/tuple_tfunc with invalid contents appearing in the Tuple parameters after intersections (such as Int, tuple of Symbol, etc).

Also sharpen valid_as_lattice / instanceof_tfunc to be able to filter out and reject types that cannot appear as tags at runtime, except where they are used for non-tag queries (like fieldtype_tfunc and subtype_tfunc).

Fixes #51311
Fixes #51228 (part 2)

aviatesk commented on base/compiler/tfuncs.jl:
```diff
@@ -461,7 +461,7 @@ function sizeof_nothrow(@nospecialize(x))
                sizeof_nothrow(rewrap_unionall(xu.b, x))
     end
-    t, exact, isconcrete = instanceof_tfunc(x)
+    t, exact, isconcrete = instanceof_tfunc(x, false)
```
We probably want to convert all `instanceof_tfunc(x, false)` to `instanceof_tfunc(x)`? I'm fine with the current state if you prefer explicitly specifying the value of the `astag` argument.

vtjnash replied:
I like the more explicit nature of it

aviatesk reviewed (approved):
LGTM. While it does introduce some complexity to our codebase, mostly due to the extra arguments in `valid_as_lattice` and `instanceof_tfunc`, but it's great that we now identify invalid type parameters and can fix the issue.

vtjnash commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -485,7 +485,10 @@ function abstract_call_method(interp::AbstractInterpreter,
-    sigtuple isa DataType || return MethodCallResult(Any, false, false, nothing, Effects())
+    sigtuple isa DataType ||
+        return MethodCallResult(Any, false, false, nothing, Effects())
+    _all(@nospecialize(x) -> valid_as_lattice(unwrapva(x), true), sigtuple.parameters) ||
```
I thought CC calls it `_all`?

aviatesk replied:
We have `all` and `any` in `CC` these days: https://github.com/JuliaLang/julia/blob/14119e0e2c1b6513b0e7e363a0d7e8dd49b63462/base/compiler/utilities.jl#L31

vtjnash committed:
inference: avoid inferring unreachable code methods

Avoids causing issues in matching_cache_argtypes/tuple_tfunc with
invalid contents appearing in the Tuple parameters after intersections
(such as Int, tuple of Symbol, etc).

Also sharpen valid_as_lattice / instanceof_tfunc to be able to filter
out and reject types that cannot appear as tags at runtime, except where
they are used for non-tag queries (like fieldtype_tfunc and subtype_tfunc).

Fixes #51228 (part 2)

aviatesk commented:
We also need to resolve the conflict in `test/compiler/inference.jl`.

vtjnash committed:
inference: guard fieldtype calls more also against invalid types

vtjnash commented:
Ah oops, it became conflicted again between the time I rebased and pushed
